drivers/usb/device/usb_dc_sam0.c

```diff
@@ -0,0 +1,527 @@
+/*
+ * Copyright (c) 2018 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "usb/dc"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_USB_DRIVER_LEVEL
+#include <logging/sys_log.h>
+
+#include <drivers/usb/usb_dc.h>
+#include <soc.h>
+#include <string.h>
+
+#define NVM_USB_PAD_TRANSN_POS 45
+#define NVM_USB_PAD_TRANSN_SIZE 5
+#define NVM_USB_PAD_TRANSP_POS 50
+#define NVM_USB_PAD_TRANSP_SIZE 5
+#define NVM_USB_PAD_TRIM_POS 55
+#define NVM_USB_PAD_TRIM_SIZE 3
+
+#define USB_SAM0_IN_EP 0x80
+
+#define REGS ((Usb *)CONFIG_USB_DC_SAM0_BASE_ADDRESS)
+
+struct usb_sam0_data {
+	UsbDeviceDescriptor descriptors[USB_EPT_NUM];
```

Comment:
You could also get endpoint number from DT instead of ASF.

Reply:
Done